File: expansion-packs/bmad-research-dev/templates/servers/zotero/get-collections.ts
/**
 * Zotero Collections Server - Direct Web API
 * Makes direct HTTPS requests to api.zotero.org (NO MCP)
 *
 * API Docs: https://www.zotero.org/support/dev/web_api/v3/basics
 *
 * Usage:
 *   import { getCollections } from './servers/zotero/get-collections.ts'
 *   const collections = await getCollections()
 */

// Environment variables (set in .env file)
declare const ZOTERO_API_KEY: string;
declare const ZOTERO_USER_ID: string;

interface Collection {
  key: string;
  version: number;
  library: {
    type: string;
    id: number;
    name: string;
  };
  data: {
    key: string;
    version: number;
    name: string;
    parentCollection: boolean | string; // false or parent key
  };
  meta: {
    numCollections: number; // Number of subcollections
    numItems: number; // Number of items
  };
}

interface ZoteroItem {
  key: string;
  data: {
    itemType: string;
    title: string;
    creators: any[];
    date?: string;
    tags: any[];
  };
}

/**
 * Get environment configuration
 */
function getConfig(): { apiKey: string; userId: string } {
  const apiKey = typeof ZOTERO_API_KEY !== 'undefined' ? ZOTERO_API_KEY : '';
  const userId = typeof ZOTERO_USER_ID !== 'undefined' ? ZOTERO_USER_ID : '';

  if (!apiKey || !userId) {
    throw new Error(
      'Zotero credentials not found. Set ZOTERO_API_KEY and ZOTERO_USER_ID in .env file.'
    );
  }

  return { apiKey, userId };
}

/**
 * Make direct API request to Zotero
 */
async function apiRequest<T>(
  endpoint: string,
  params: Record<string, string> = {}
): Promise<T> {
  const { apiKey, userId } = getConfig();

  // Build URL with query parameters
  const url = new URL(`https://api.zotero.org/users/${userId}${endpoint}`);
  Object.entries(params).forEach(([key, value]) => {
    if (value) url.searchParams.append(key, value);
  });

  console.log(`[Zotero API] GET ${url.pathname}${url.search}`);

  // Make HTTPS request
  const response = await fetch(url.toString(), {
    method: 'GET',
    headers: {
      'Zotero-API-Key': apiKey,
      'Zotero-API-Version': '3',
    },
  });

  if (!response.ok) {
    throw new Error(
      `Zotero API error: ${response.status} ${response.statusText}`
    );
  }

  return (await response.json()) as T;
}

/**
 * Get all collections in your Zotero library
 *
 * @returns Array of collections with hierarchy information
 *
 * @example
 * const collections = await getCollections()
 * collections.forEach(c => {
 *   console.log(`${c.data.name}: ${c.meta.numItems} items`)
 * })
 */
export async function getCollections(): Promise<Collection[]> {
  console.log(`[Zotero] Getting all collections`);

  const collections = await apiRequest<Collection[]>('/collections', {
    limit: '100',
  });

  console.log(`[Zotero] Found ${collections.length} collections`);

  return collections;
}

/**
 * Get items in a specific collection
 */
export async function getCollectionItems(
  collectionKey: string,
  limit?: number
): Promise<ZoteroItem[]> {
  console.log(`[Zotero] Getting items in collection: ${collectionKey}`);

  const items = await apiRequest<ZoteroItem[]>(
    `/collections/${collectionKey}/items`,
    {
      limit: (limit || 100).toString(),
    }
  );

  console.log(`[Zotero] Found ${items.length} items in collection`);

  return items;
}

/**
 * Find collection by name (case-insensitive)
 */
export async function findCollectionByName(
  name: string
): Promise<Collection | undefined> {
  const collections = await getCollections();

  return collections.find((c) =>
    c.data.name.toLowerCase().includes(name.toLowerCase())
  );
}

/**
 * Get collection hierarchy (parent-child relationships)
 */
export async function getCollectionHierarchy(): Promise<{
  topLevel: Collection[];
  byParent: Record<string, Collection[]>;
}> {
  const collections = await getCollections();

  const topLevel: Collection[] = [];
  const byParent: Record<string, Collection[]> = {};

  collections.forEach((collection) => {
    if (
      collection.data.parentCollection === false ||
      !collection.data.parentCollection
    ) {
      topLevel.push(collection);
    } else {
      const parentKey =
        typeof collection.data.parentCollection === 'string'
          ? collection.data.parentCollection
          : '';
      if (!byParent[parentKey]) {
        byParent[parentKey] = [];
      }
      byParent[parentKey].push(collection);
    }
  });

  return { topLevel, byParent };
}

/**
 * Print collection hierarchy in tree format
 */
export async function printCollectionHierarchy(): Promise<void> {
  const { topLevel, byParent } = await getCollectionHierarchy();

  console.log(`## Your Zotero Collections\n`);

  function printCollection(collection: Collection, indent: number = 0) {
    const prefix = '  '.repeat(indent);
    console.log(
      `${prefix}📁 ${collection.data.name} (${collection.meta.numItems} items)`
    );

    // Print children
    const children = byParent[collection.key] || [];
    children.forEach((child) => printCollection(child, indent + 1));
  }

  topLevel.forEach((collection) => printCollection(collection));
}

/**
 * Get total item count across all collections
 */
export async function getTotalItemCount(): Promise<{
  total: number;
  byCollection: Record<string, number>;
}> {
  const collections = await getCollections();

  const byCollection: Record<string, number> = {};
  let total = 0;

  collections.forEach((collection) => {
    byCollection[collection.data.name] = collection.meta.numItems;
    total += collection.meta.numItems;
  });

  return { total, byCollection };
}

/**
 * Find collections by keyword
 */
export async function findCollectionsByKeyword(
  keyword: string
): Promise<Collection[]> {
  const collections = await getCollections();

  return collections.filter((c) =>
    c.data.name.toLowerCase().includes(keyword.toLowerCase())
  );
}

/**
 * Get largest collections
 */
export async function getLargestCollections(
  limit: number = 10
): Promise<Collection[]> {
  const collections = await getCollections();

  return collections
    .sort((a, b) => b.meta.numItems - a.meta.numItems)
    .slice(0, limit);
}

File: expansion-packs/bmad-research-dev/templates/servers/zotero/get_collections.py
"""
Zotero Collections Server - Direct Web API
Makes direct HTTPS requests to api.zotero.org (NO MCP)

API Docs: https://www.zotero.org/support/dev/web_api/v3/basics

Usage:
    from servers.zotero.get_collections import get_collections
    collections = get_collections()
"""

import os
import requests
from typing import List, Dict, Optional, Any, Union
from dotenv import load_dotenv

# Load environment variables
load_dotenv()


def get_config() -> Dict[str, str]:
    """Get environment configuration"""
    api_key = os.getenv('ZOTERO_API_KEY', '')
    user_id = os.getenv('ZOTERO_USER_ID', '')

    if not api_key or not user_id:
        raise ValueError(
            'Zotero credentials not found. Set ZOTERO_API_KEY and ZOTERO_USER_ID in .env file.'
        )

    return {'apiKey': api_key, 'userId': user_id}


def api_request(endpoint: str, params: Optional[Dict[str, str]] = None) -> Any:
    """Make direct API request to Zotero"""
    config = get_config()

    if params is None:
        params = {}

    # Build URL
    url = f"https://api.zotero.org/users/{config['userId']}{endpoint}"

    print(f"[Zotero API] GET {endpoint}")

    # Make HTTPS request
    headers = {
        'Zotero-API-Key': config['apiKey'],
        'Zotero-API-Version': '3'
    }

    response = requests.get(url, headers=headers, params=params)

    if not response.ok:
        raise Exception(f"Zotero API error: {response.status_code} {response.reason}")

    return response.json()


def get_collections() -> List[Dict[str, Any]]:
    """
    Get all collections in your Zotero library

    Returns:
        Array of collections with hierarchy information

    Example:
        collections = get_collections()
        for c in collections:
            print(f"{c['data']['name']}: {c['meta']['numItems']} items")
    """
    print("[Zotero] Getting all collections")

    collections = api_request('/collections', {'limit': '100'})

    print(f"[Zotero] Found {len(collections)} collections")

    return collections


def get_collection_items(collection_key: str, limit: int = 100) -> List[Dict[str, Any]]:
    """Get items in a specific collection"""
    print(f"[Zotero] Getting items in collection: {collection_key}")

    items = api_request(f'/collections/{collection_key}/items', {'limit': str(limit)})

    print(f"[Zotero] Found {len(items)} items in collection")

    return items


def find_collection_by_name(name: str) -> Optional[Dict[str, Any]]:
    """Find collection by name (case-insensitive)"""
    collections = get_collections()

    for c in collections:
        if name.lower() in c['data']['name'].lower():
            return c

    return None


def get_collection_hierarchy() -> Dict[str, Any]:
    """Get collection hierarchy (parent-child relationships)"""
    collections = get_collections()

    top_level = []
    by_parent: Dict[str, List[Dict[str, Any]]] = {}

    for collection in collections:
        parent = collection['data'].get('parentCollection')
        if parent is False or not parent:
            top_level.append(collection)
        else:
            parent_key = parent if isinstance(parent, str) else ''
            if parent_key not in by_parent:
                by_parent[parent_key] = []
            by_parent[parent_key].append(collection)

    return {
        'topLevel': top_level,
        'byParent': by_parent
    }


def print_collection_hierarchy() -> None:
    """Print collection hierarchy in tree format"""
    hierarchy = get_collection_hierarchy()

    print("\n## Your Zotero Collections\n")

    def print_collection(collection: Dict[str, Any], indent: int = 0):
        prefix = '  ' * indent
        name = collection['data']['name']
        num_items = collection['meta']['numItems']
        print(f"{prefix}📁 {name} ({num_items} items)")

        # Print children
        children = hierarchy['byParent'].get(collection['key'], [])
        for child in children:
            print_collection(child, indent + 1)

    for collection in hierarchy['topLevel']:
        print_collection(collection)


def get_total_item_count() -> Dict[str, Any]:
    """Get total item count across all collections"""
    collections = get_collections()

    by_collection: Dict[str, int] = {}
    total = 0

    for collection in collections:
        name = collection['data']['name']
        num_items = collection['meta']['numItems']
        by_collection[name] = num_items
        total += num_items

    return {
        'total': total,
        'byCollection': by_collection
    }


def find_collections_by_keyword(keyword: str) -> List[Dict[str, Any]]:
    """Find collections by keyword"""
    collections = get_collections()

    return [
        c for c in collections
        if keyword.lower() in c['data']['name'].lower()
    ]


def get_largest_collections(limit: int = 10) -> List[Dict[str, Any]]:
    """Get largest collections"""
    collections = get_collections()

    sorted_collections = sorted(
        collections,
        key=lambda c: c['meta']['numItems'],
        reverse=True
    )

    return sorted_collections[:limit]
